crates/assists/src/handlers/infer_function_return_type.rs

```diff
@@ -0,0 +1,237 @@
+                ast::Expr::BlockExpr(block) => (block.expr()?, false),
+                body => (body, true),
+            };
+            let ret_ty = closure.ret_type();
+            let rpipe = closure.param_list()?.syntax().last_token()?;
+            (ret_ty, tail_expr, rpipe, wrap_expr)
+        } else {
+            let func = expr.syntax().ancestors().find_map(ast::Fn::cast)?;
+            let tail_expr = func.body()?.expr()?;
+            let ret_ty = func.ret_type();
+            let rparen = func.param_list()?.r_paren_token()?;
+            (ret_ty, tail_expr, rparen, false)
+        };
+    if ret_ty.is_some() {
+        mark::hit!(existing_ret_type);
+        mark::hit!(existing_ret_type_closure);
+        return None;
+    }
+    // check whether the expr we were at is indeed the tail expression
+    if !tail_expr.syntax().text_range().contains_range(expr.syntax().text_range()) {
+        mark::hit!(not_tail_expr);
+        return None;
+    }
+    Some((tail_expr, insert_pos, wrap_expr))
+}
+
+#[cfg(test)]
+mod tests {
+    use crate::tests::{check_assist, check_assist_not_applicable};
+
+    use super::*;
+
+    #[test]
+    fn infer_return_type() {
+        check_assist(
+            infer_function_return_type,
+            r#"fn foo() {
+                45<|>
+            }"#,
+            r#"fn foo() -> i32 {
+                45
+            }"#,
+        );
+    }
+
+    #[test]
+    fn infer_return_type_nested() {
+        check_assist(
+            infer_function_return_type,
+            r#"fn foo() {
+                if true {
+                    3<|>
+                } else {
+                    5
+                }
+            }"#,
+            r#"fn foo() -> i32 {
+                if true {
+                    3
+                } else {
+                    5
+                }
+            }"#,
+        );
+    }
+
+    #[test]
+    fn not_applicable_ret_type_specified() {
+        mark::check!(existing_ret_type);
+        check_assist_not_applicable(
+            infer_function_return_type,
+            r#"fn foo() -> i32 {
+                ( 45<|> + 32 ) * 123
+            }"#,
+        );
+    }
+
+    #[test]
+    fn not_applicable_non_tail_expr() {
+        mark::check!(not_tail_expr);
+        check_assist_not_applicable(
+            infer_function_return_type,
+            r#"fn foo() {
+                let x = <|>3;
+                ( 45 + 32 ) * 123
+            }"#,
+        );
+    }
+
+    #[test]
+    fn not_applicable_unit_return_type() {
+        check_assist_not_applicable(
+            infer_function_return_type,
+            r#"fn foo() {
+                (<|>)
+            }"#,
+        );
+    }
+
+    #[test]
+    fn infer_return_type_closure_block() {
+        check_assist(
+            infer_function_return_type,
+            r#"fn foo() {
+                |x: i32| {
+                    x<|>
+                };
+            }"#,
+            r#"fn foo() {
```

Comment:
Preferred way to format text is without indentation:

https://github.com/rust-analyzer/rust-analyzer/commit/0fc8fd2bd589572964d4949cd35aac44659af334

That way, it doesn't look weird if printed, and the offsets inside slection match offsets in the fixture.